cranelift/frontend/src/ssa.rs

```diff
@@ -405,7 +359,70 @@ impl SSABuilder {
     /// Note: use only when you know what you are doing, this might break the SSA building problem
     pub fn remove_block_predecessor(&mut self, block: Block, inst: Inst) -> Block {
         debug_assert!(!self.is_sealed(block));
-        self.ssa_blocks[block].remove_predecessor(inst)
+        let pred = self.ssa_blocks[block].remove_predecessor(inst);
+        self.update_predecessor_cycle(block, pred, false);
+        pred
+    }
+
+    /// Call this when `pred` is not in the predecessors of `block`: before adding, or after
+    /// removing.
+    fn update_predecessor_cycle(&mut self, block: Block, mut pred: Block, add: bool) {
```

Comment:
Can we make `add` a small helper enum type, like `enum AddRemove { Add, Remove }` ? That would make the meaning of the bool much clearer.